modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/SqlPlanHistoryIntegrationTest.java

```diff
@@ -0,0 +1,718 @@
+            }
+            catch (Exception ignore) {
+                //No-Op
+            }
+
+            checkSqlPlanHistory((!isClient && sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME) ? 1 : 0);
+        });
+    }
+
+    /** Checks successful SqlQuery. */
+    @Test
+    public void testSqlQuery() {
+        runSuccessfulQuery(sqlQry);
+    }
+
+    /** Checks failed SqlQuery. */
+    @Test
+    public void testSqlQueryFailed() {
+        runFailedQuery(sqlQryFailed);
+    }
+
+    /** Checks ScanQuery. */
+    @Test
+    public void testScanQuery() {
+        runQueryWithoutPlan(scanQry);
+    }
+
+    /** Checks TextQuery. */
+    @Test
+    public void testTextQuery() {
+        runQueryWithoutPlan(textQry);
+    }
+
+    /** Checks DML commands executed via JDBC. */
+    @Test
+    public void testJdbcDml() {
+        runJdbcDml(dmlCmds);
+    }
+
+    /** Checks DML commands with joins executed via JDBC. */
+    @Test
+    public void testJdbcDmlWithJoins() {
+        runJdbcDml(dmlCmdsWithJoins);
+    }
+
+    /** Checks DML commands executed via SqlFieldsQuery. */
+    @Test
+    public void testSqlFieldsDml() {
+        runSqlFieldsDml(dmlCmds);
+    }
+
+    /** Checks DML commands with joins executed via SqlFieldsQuery. */
+    @Test
+    public void testSqlFieldsDmlWithJoins() {
+        runSqlFieldsDml(dmlCmdsWithJoins);
+    }
+
+    /** Checks that older plan entries are evicted when maximum history size is reached. */
+    @Test
+    public void testPlanHistoryEviction() throws Exception {
+        assumeFalse("No SQL plans are written on the client node when using the H2 engine",
+            isClient && sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME);
+
+        for (int i = 1; i <= (PLAN_HISTORY_SIZE + PLAN_HISTORY_EXCESS); i++) {
+            try {
+                SqlFieldsQuery qry = new SqlFieldsQuery(SQL + " where A.fail()=" + i);
+
+                cacheQuery(qry.setLocal(loc), "A");
+            }
+            catch (Exception ignore) {
+                //No-op
+            }
+        }
+
+        GridTestUtils.waitForCondition(() -> getSqlPlanHistory().size() == PLAN_HISTORY_SIZE, 1000);
+
+        checkSqlPlanHistory(PLAN_HISTORY_SIZE);
+
+        Set<String> qrys = getSqlPlanHistory().keySet().stream()
+            .map(SqlPlan::query)
+            .collect(Collectors.toSet());
+
+        for (int i = 1; i <= PLAN_HISTORY_EXCESS; i++)
+            assertFalse(qrys.contains(SQL + " where A.fail=" + i));
+    }
+
+    /**
+     * Checks that older SQL plan history entries are replaced with newer ones with the same parameters (except for
+     * the beginning time).
+     */
+    @Test
+    public void testEntryReplacement() throws IgniteInterruptedCheckedException {
+        assumeFalse("With the H2 engine, scan counts can be added to SQL plans for local queries ",
+            loc && sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME);
```

Comment:
The “scanCount: X” suffix in the SQL plan shows how many data pages were scanned while the query was running. It helps us understand how well the query performs. A high scanCount usually means that a large portion of the table was read, which might indicate that the query isn't very efficient or that certain indexes are missing.

The scanCount metric is only visible when using the local H2 engine, which runs queries on a single node. For these queries, H2 provides a more detailed, node-specific execution plan, including metrics like scanCount. As for distributed queries, execution plans are combined and shown at a higher level, without exposing H2’s internal metrics specific to local execution.

As we can see from the tests, the scanCount metric appears only after running the same local query several times. This happens because when we execute a SQL query for the first time, the H2 engine compiles an execution plan from scratch. During this initial execution, the plan is generated, but runtime statistics such as scanCount are not immediately included in the plan output because they are collected during or after execution.

The plan is then cached, and subsequent executions of the same query reuse this cached plan. At this point, the engine incorporates metrics (such as scanCount) from previous executions.

Keeping that in mind, it would seem more logical to replace the local H2 queries’ plans without a scanCount with those containing a scanCount since they potentially provide more detailed information about query execution.